### PR TITLE
Batch prover

### PR DESCRIPTION
**Pull request by hadasz**

### CODE REVIEW

File: uni-stark/src/prover.rs [hunk 1]
```diff
@@ -83,32 +129,134 @@ where
...
+        challenger,
+        &trace_data,
+    );
+
+    challenger.observe(quotient_commit.clone());
+
+    ((trace_commit, trace_data), (quotient_commit, quotient_data))
+}
+
+    #[instrument(skip_all)]
```
Comment (morganthomas) [inline uni-stark/src/prover.rs:217]: Looks like an indentation error here and line 219?

File: uni-stark/src/prover.rs [hunk 2]
```diff
@@ -83,32 +129,134 @@ where
                 pcs.coset_shift().exp_power_of_2(log_quotient_degree),
             )
         });
+    (quotient_commit, quotient_data)
+}
+
+pub fn get_trace_lde<SC>(config: &SC, trace: RowMajorMatrix<SC::Val>) -> Commitment<SC>
```
Comment (morganthomas) [inline uni-stark/src/prover.rs:135]: What is a trace LDE?

Comment (morganthomas) [reply uni-stark/src/prover.rs:135]: "Trace low degree extension" I suppose.

**Review by morganthomas — APPROVED** [review]

File: commit/src/adapters/extension_mmcs.rs [hunk 3]
```diff
@@ -34,6 +34,9 @@ where
     type Error = InnerMmcs::Error;
     type Mat<'a> = ExtensionMatrix<F, EF, InnerMmcs::Mat<'a>> where Self: 'a;
 
+    fn combine(&self, data:&Vec<Self::ProverData>) -> (Self::Commitment, Self::ProverData){
```
Comment (morganthomas) [inline commit/src/adapters/extension_mmcs.rs:37]: This produces:

```
warning: function cannot return without recursing
  --> commit/src/adapters/extension_mmcs.rs:37:5
   |
37 |     fn combine(&self, data:&Vec<Self::ProverData>) -> (Self::Commitment, Self::ProverData){
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
38 |     self.combine(data)
   |     ------------------ recursive call site
   |
   = help: a `loop` may express intention better if this is on purpose
   = note: `#[warn(unconditional_recursion)]` on by default
```

File: ldt/src/quotient.rs [hunk 4]
```diff
@@ -157,6 +157,10 @@ where
     type Error = QuotientError<Inner::Error>;
     type Mat<'a> = QuotientMatrix<F, EF, Inner::Mat<'a>> where Self: 'a;
 
+    fn combine(&self, data:&Vec<Self::ProverData>) -> (Self::Commitment, Self::ProverData) {
+	self.combine(data)
```
Comment (morganthomas) [inline ldt/src/quotient.rs:161]: This produces:
```
warning: function cannot return without recursing
   --> ldt/src/quotient.rs:160:5
    |
160 |     fn combine(&self, data:&Vec<Self::ProverData>) -> (Self::Commitment, Self::ProverData) {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
161 |     self.combine(data)
    |     ------------------ recursive call site
    |
    = help: a `loop` may express intention better if this is on purpose
    = note: `#[warn(unconditional_recursion)]` on by default
```

File: uni-stark/src/prover.rs [hunk 5]
```diff
@@ -9,63 +8,110 @@ use p3_field::{
...
-#[instrument(skip_all)]
-pub fn prove<
-    SC,
-    #[cfg(debug_assertions)] A: for<'a> Air<crate::check_constraints::DebugConstraintBuilder<'a, SC::Val>>,
-    #[cfg(not(debug_assertions))] A,
->(
+
+pub fn open<SC>(
     config: &SC,
-    air: &A,
```
Comment (morganthomas) [inline uni-stark/src/prover.rs:31]: Is there a reason there are not type synonyms for these rather clunky type signatures in this function signature, namely for `trace_data`, `quotient_data`, and the return value?

File: uni-stark/src/prover.rs (hunk 1 above)
Comment (hadasz) [reply uni-stark/src/prover.rs]: I think this was fixed by latest commit

File: uni-stark/src/prover.rs (hunk 2 above)
Comment (hadasz) [reply uni-stark/src/prover.rs:132]: Yes

File: commit/src/adapters/extension_mmcs.rs (hunk 3 above)
Comment (hadasz) [reply commit/src/adapters/extension_mmcs.rs]: I think hopefully this is fixed by latest commit

File: ldt/src/quotient.rs (hunk 4 above)
Comment (hadasz) [reply ldt/src/quotient.rs]: Fixed by latest commit

File: uni-stark/src/prover.rs (hunk 5 above)
Comment (hadasz) [reply uni-stark/src/prover.rs:31]: I have added type synonyms for ProverData and Commitment so function signatures should be clearer now

**Review by morganthomas — APPROVED** [review]

**Comment by dlubarov** [comment]
We ended up not needing `combine` in the end, and ended up not using those bits of uni-stark code since they didn't support permutation traces

**Pull request closed without merging**